add support for node-selector

dschiemann80 opened the pull request:
- introduce cmdline arg and docker label for setting a node selector
- format is a comma-separated list of key-value pairs

I tried to conform to kubedock conventions as best as I could. I wasn't sure where to put the parsing code, so I went with containers.go as a first try.
Please let me know if you find this useful and if I should adapt anything.

joyrex2001 commented:
Thanks appreciated, sorry for the late response.

I aggree with the suggested format, and that it should be configureable via a label as well as cli arguments. Note that the applying a node selector should also be possible with the use of a pod-template. However, in this case it can be dynamically applied (by using the labels) as well, which is a nice feature to have.

Generally, if such a label is introduced, the processing of the label is handled in `internal/model/types/container.go` and the routes implementations (`docker/containers.go` and `libpod/containers.go`) make sure the label is present (as you already did for the docker variant). For example, see how the pull policy is implemented, where the label is populated, and the value is processed and retrieved in the model implementation with GetImagePullPolicy. The node-selector should be implemented in a similar fashion.

Can you move the code to the model, and add some unit tests as well?
Can you also update the documentation in the `README.md` and `config.md` files?

dschiemann80 commented:
Thanks for your feedback.

I moved the code to the model as you suggested and added the label handling to the libpod routes. I am now honoring the node selector from the pod template by passing it into GetNodeSelector() as an argument.
I also added unit tests and documentation.

Please let me know if you would like to have adapted anything else.

joyrex2001 commented on internal/model/types/container.go:
```diff
@@ -37,6 +37,7 @@ type Container struct {
 	MappedPorts    map[int]int
 	Networks       map[string]interface{}
 	NetworkAliases []string
+	NodeSelector   map[string]string
```
This isn't used anymore.

joyrex2001 commented on internal/server/routes/docker/containers.go:
```diff
@@ -91,6 +94,7 @@ func ContainerCreate(cr *common.ContextRouter, c *gin.Context) {
 		Binds:        in.HostConfig.Binds,
 		Mounts:       mounts,
 		PreArchives:  []types.PreArchive{},
+		NodeSelector: map[string]string{},
```
Same, this isn't used anymore.

joyrex2001 commented on internal/server/routes/libpod/containers.go:
```diff
@@ -74,6 +77,7 @@ func ContainerCreate(cr *common.ContextRouter, c *gin.Context) {
 		ExposedPorts: map[string]interface{}{},
 		ImagePorts:   map[string]interface{}{},
 		Labels:       in.Labels,
+		NodeSelector: map[string]string{},
```
And this one as well.

joyrex2001 reviewed (changes requested):
Can you remove the unused code?

dschiemann80 commented:
Sure, removed it.

joyrex2001 commented:
Awesome, thanks @dschiemann80 !
